cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -41,6 +41,84 @@ namespace internal {
 using ::arrow::internal::make_unique;
 }  // namespace internal
 
+Status DecodeArg(const substrait::FunctionArgument& arg, int idx, SubstraitCall* call,
```

Comment:
does `int idx` needs to be `uint32_t idx`?

Reply:
I think I was originally aiming for compatibility with the source, which is a Substrait (i.e. protobuf) index which is signed.  But I agree it looks nicer with `uint32_t` since that matches the destination (and it seems I was inconsistent with casting).  I have updated it.